### ISSUE TITLE
@contextmanager not applied to overloaded functions

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#3619]
**Describe the bug**
It seems that the `@contextmanager` is not applied to overloaded functions (mypy behaves the same as pyright).

**To Reproduce**
Without overloads, no errors are reported:

```py
# pyright: strict
from __future__ import annotations

from contextlib import contextmanager
from io import (
    BytesIO,
    StringIO,
)
from typing import Iterator

@contextmanager
def get_buffer(buf: str | None) -> Iterator[BytesIO] | Iterator[StringIO]:
    ...

with get_buffer("") as foo:
    reveal_type(foo)  # Type of "foo" is "BytesIO | StringIO"
```
But when using overloads, errors are reported:

```py
# pyright: strict
from __future__ import annotations

from contextlib import contextmanager
from io import (
    BytesIO,
    StringIO,
)
from typing import (
    Iterator,
    overload,
)

@overload
def get_buffer(buf: str) -> Iterator[BytesIO]:
    ...

@overload
def get_buffer(buf: None) -> Iterator[StringIO]:
    ...

@contextmanager
def get_buffer(buf: str | None) -> Iterator[BytesIO] | Iterator[StringIO]:
    ...

with get_buffer("") as foo:   # error: Object of type "Iterator[BytesIO]" cannot be used with "with" because it does not implement __enter__ (and the same for __exit__)
    reveal_type(foo)  # Type of "foo" is "Unknown"

```
**Expected behavior**
I would expect that  `@overload` provides type information for the original function before it is decorated.

**Comment by erictraut** [comment]
The behavior is correct. You are applying the `@contextmanager` only to the implementation, not the overloads. 

If this is your internal code and you are using only pyright as a type checker, you can use the following approach:

```python
@overload
@contextmanager
def get_buffer(buf: str) -> Iterator[BytesIO]:
    ...

@overload
@contextmanager
def get_buffer(buf: None) -> Iterator[StringIO]:
    ...
```

This may not work with other type checkers, however, so I'd recommend the following approach instead:

```python
@overload
def get_buffer(buf: str) -> AbstractContextManager[Iterator[BytesIO]]:
    ...

@overload
def get_buffer(buf: None) -> AbstractContextManager[Iterator[StringIO]]:
    ...
```

**Comment by twoertwein** [comment]
> The behavior is correct. You are applying the `@contextmanager` only to the implementation, not the overloads.

Wow, thank you! That is really good to know!
